test-data/unit/check-protocols.test

```diff
@@ -0,0 +1,1798 @@
+x3 = y3 # E: Incompatible types in assignment (expression has type Pinv[A], variable has type Pinv[B])
+[out]
+
+[case testGenericProtocolsInference1]
+from typing import Protocol, Sequence, TypeVar
+
+T = TypeVar('T', covariant=True)
+
+class Closeable(Protocol[T]):
+    def close(self) -> T:
+        pass
+
+class F:
+    def close(self) -> int:
+        return 0
+
+def close(arg: Closeable[T]) -> T:
+    return arg.close()
+
+def close_all(args: Sequence[Closeable[T]]) -> T:
+    for arg in args:
+            arg.close()
+    return args[0].close()
+
+arg: Closeable[int]
+
+reveal_type(close(F())) # E: Revealed type is 'builtins.int*'
+reveal_type(close(arg)) # E: Revealed type is 'builtins.int*'
+reveal_type(close_all([F()])) # E: Revealed type is 'builtins.int*'
+reveal_type(close_all([arg])) # E: Revealed type is 'builtins.int*'
+[builtins fixtures/isinstancelist.pyi]
+[out]
+
+[case testProtocolGenericInference2]
+from typing import Generic, TypeVar, Protocol
+T = TypeVar('T')
+S = TypeVar('S')
+
+class P(Protocol[T, S]):
+    x: T
+    y: S
+
+class C:
+    x: int
+    y: int
+
+def fun3(x: P[T, T]) -> T:
+    pass
+reveal_type(fun3(C())) # E: Revealed type is 'builtins.int*'
+[out]
+
+[case testProtocolGenericInferenceCovariant]
+from typing import Generic, TypeVar, Protocol
+T = TypeVar('T', covariant=True)
+S = TypeVar('S', covariant=True)
+U = TypeVar('U')
+
+class P(Protocol[T, S]):
+    def x(self) -> T: pass
+    def y(self) -> S: pass
+
+class C:
+    def x(self) -> int: pass
+    def y(self) -> int: pass
+
+def fun4(x: U, y: P[U, U]) -> U:
+    pass
+reveal_type(fun4('a', C())) # E: Revealed type is 'builtins.object*'
+[out]
+
+[case testGenericSubProtocols]
+from typing import TypeVar, Protocol, Tuple, Generic
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+class P1(Protocol[T]):
+    attr1: T
+class P2(P1[T], Protocol[T, S]):
+    attr2: Tuple[T, S]
+
+class C:
+    def __init__(self, a1: int, a2: Tuple[int, int]) -> None:
+        self.attr1 = a1
+        self.attr2 = a2
+
+c: C
+var: P2[int, int] = c
+var2: P2[int, str] = c # E: Incompatible types in assignment (expression has type "C", variable has type P2[int, str]) \
+                       # N: Following member(s) of "C" have conflicts: \
+                       # N:     attr2: expected "Tuple[int, str]", got "Tuple[int, int]"
+
+class D(Generic[T]):
+    attr1: T
+class E(D[T]):
+    attr2: Tuple[T, T]
+
+def f(x: T) -> T:
+    z: P2[T, T] = E[T]()
+    y: P2[T, T] = D[T]() # E: Incompatible types in assignment (expression has type D[T], variable has type P2[T, T]) \
+                         # N: 'D' is missing following 'P2' protocol member: \
+                         # N:     attr2
+    return x
+[builtins fixtures/isinstancelist.pyi]
+[out]
+
+[case testGenericSubProtocolsExtensionInvariant]
+from typing import TypeVar, Protocol, Union
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+class P1(Protocol[T]):
```

Comment:
Do you test somewhere if two generic protocols with an equivalent structure are compatible with each other?